_blog/2024-01-23-using-serde-rust.mdx

```diff
@@ -0,0 +1,341 @@
+    }
+}
+```
+
+There is also documentation on deserializing structs which you can find [here.](https://serde.rs/deserialize-struct.html) However, generally speaking it is recommended that you use the `derive` feature macros as the manual implementation (as seen on the page itself) is quite large. The implementation involves mostly using a visitor that can visit a map or sequence then iterate through the elements to deserialize it. 
+
+### Using serde attributes
+
+When it comes to serde, the crate also has a number of useful attribute macros that we can use on our types to allow things like field renaming when deserializing a field or serializing to a struct. One of the best examples of this would be when you’re interacting with an API written in a language that may have a key that is a reserved keyword in Rust. You can add a `#[serde(rename)]` attribute macro like so:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct MyStruct {
+    #[serde(rename = "type")]
+    kind: String
+}
+```
+
+This allows you to get around the issue!
+
+You can also rename all of your fields to another casing by using the `rename_all` attribute:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MyStruct {
+    my_message: String
+}
+```
+
+Now when you serialize this struct, `my_message` should automatically turn into `myMessage`! Perfect for working with APIs written in other languages or with different conventions.
+
+If you’d prefer to not wrap fields in `Option`, you can also implement default values by using `#[serde(default)]`. This simply allows fields to be filled in with default values instead of automatically erroring out. You can also use `#[serde(default = "path")]` to be able to point to functions for providing the automatic default. For example, this struct and function:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct MyStruct {
+	#[serde(path = "my_function")]
+    my_message: String,
+}
+
+fn my_function() -> String {
+    "Hello world!".to_string()
+}
+```
+
+`serde` also offers other useful attributes, like being able to deny unknown fields using `#[serde(deny_unknown_fields)]` on top of the struct. This allows you to make sure that the struct is exactly as-is when serializing and deserializing. 
+
+### Deserializing and Serializing enums
+
+Let’s examine this enum type:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+enum MyEnum {
+	Data { id: String, data: Value },
+    SomeOtherData { id: i32, name: String }
+}
+```
+
+Note that when converting to and from this enumstruct, it can take two options:
+
+- A String field named “id” and a JSON value with the key “data” (this can be a map, a value or anything that the `Json` value can hold)
+- An `i32` field named “id” and a `String` field named “name”
```

Comment:
```suggestion
- A String field named `id` and a JSON value with the key `data` (this can be a map, a value or anything that the `Json` value can hold).
- An `i32` field named `id` and a `String` field named `name`.
```